packages/babel-plugin-transform-object-rest-spread/src/index.js

```diff
@@ -1,17 +1,195 @@
 export default function ({ types: t }) {
+  function hasRestProperty(node) {
+    for (let property of (node.properties)) {
+      if (t.isRestProperty(property)) {
+        return true;
+      }
+    }
+  }
+
+  function variableDeclarationHasRestProperty(node) {
+    for (let declar of (node.declarations)) {
+      if (t.isObjectPattern(declar.id)) {
+        return hasRestProperty(declar.id);
+      }
+    }
+    return false;
+  }
+
   function hasSpread(node) {
-    for (let prop of (node.properties: Array<Object>)) {
+    for (let prop of (node.properties)) {
       if (t.isSpreadProperty(prop)) {
         return true;
       }
     }
     return false;
   }
 
+  function createObjectSpread(file, props, objRef) {
+    const restProperty = props.pop();
+
+    let keys = [];
+    for (let prop of props) {
+      let key = prop.key;
+      if (t.isIdentifier(key) && !prop.computed) {
+        key = t.stringLiteral(prop.key.name);
+      }
+      keys.push(key);
+    }
+
+    return t.variableDeclarator(
+      restProperty.argument,
+      t.callExpression(
+        file.addHelper("objectWithoutProperties"), [
+          objRef,
+          t.arrayExpression(keys)
+        ]
+      )
+    );
+  }
+
+  function replaceRestProperty(paramsPath, i, numParams) {
+    if (paramsPath.isObjectPattern() && hasRestProperty(paramsPath.node)) {
+      let parentPath = paramsPath.parentPath;
+      let uid = parentPath.scope.generateUidIdentifier("ref");
+
+      let declar = t.variableDeclaration("let", [
+        t.variableDeclarator(paramsPath.node, uid)
+      ]);
+      declar._blockHoist = i ? numParams - i : 1;
+
+      parentPath.ensureBlock();
+      parentPath.get("body").unshiftContainer("body", declar);
+      paramsPath.replaceWith(uid);
+    }
+  }
+
   return {
     inherits: require("babel-plugin-syntax-object-rest-spread"),
 
     visitor: {
+      // taken from transform-es2015-parameters/src/destructuring.js
+      // function a({ b, ...c }) {}
+      Function(path) {
+        let params = path.get("params");
+        for (let i = 0; i < params.length; i++) {
+          replaceRestProperty(params[i], i, params.length);
+        }
+      },
+      // adapted from transform-es2015-destructuring/src/index.js#pushObjectRest
+      // const { a, ...b } = c;
+      VariableDeclarator(path, file) {
+        if (!path.get("id").isObjectPattern()) { return; }
+        const kind = path.parentPath.node.kind;
+        let nodes = [];
+
+        path.traverse({
+          RestProperty(path) {
+            if (!this.originalPath.node) {
+              return;
+            }
```

Comment:
I think it was because at the time I was getting an infinite loop or the node was removed and I was trying to use it - will remove.